### ISSUE TITLE
Memory leak with non-tiny method name

### ISSUE THREAD

**Issue by jskeet** [grpc/grpc#11434]
Grpc version: 1.3.6 (but affects at least 1.3.0; haven't tested earlier versions yet)
Language: C#
OS: Windows

**TL;DR**

When calling a method from the C# client side, there's a small memory leak on each request (at least, the leak size seems to correlate with the number of requests made) if the method being called has "long" (anything non-tiny) fully-qualified name.

**Repro**

Consider this service:

```text
syntax = "proto3";
package test;
message Empty { }
service Serv {
  rpc M(Empty) returns (Empty);
  rpc Method(Empty) returns (Empty);
}
```

The [attached zip file](https://github.com/grpc/grpc/files/1060159/GrpcLeak.zip) has everything required to reproduce on Windows, with .NET Core installed:

- Unpack and go into the GrpcLeak directory
- Run `dotnet restore`
- Run `dotnet run -- server`
- In another window, run `dotnet run -- goodclient`
- In another window, run `dotnet run -- badclient`

The "goodclient" runs `test.Serv.M`; the "badclient" runs `test.Serv.Method`, against the locally-running (but separate process) server. Every 5 seconds, each client reports counters for requests (sent, complete, errors) and the private bytes of the process.

You'll see goodclient stay reasonably stable (about 20Mb on my machine) whereas badclient leaks memory fairly rapidly (about 1MB per 5 seconds, on average). Looking at this in Visual Studio, it looks like it's a leak of unmanaged rather than managed memory.

Currently I've only tested this in C#. If the bug turns out to be in C# client code, it would be worth writing tests in other languages too.

**Comment by jskeet** [comment]
Okay, I've made a bit of progress understanding where this goes wrong...

The method name ends up in the native [`grpcsharp_channel_create_call`](https://github.com/grpc/grpc/blob/6b19d3b408903a9a91727d1dfd4cd1704433cfe7/src/csharp/ext/grpc_csharp_ext.c#L405) which has:

```
grpc_slice method_slice = grpc_slice_from_copied_string(method);
```

Then `grpc_slice_from_copied_string` calls [`grpc_slice_from_copied_buffer`](https://github.com/grpc/grpc/blob/6b19d3b408903a9a91727d1dfd4cd1704433cfe7/src/core/lib/slice/slice.c#L206) which uses `GRPC_SLICE_MALLOC`.

That's [defined](https://github.com/grpc/grpc/blob/6b19d3b408903a9a91727d1dfd4cd1704433cfe7/include/grpc/slice.h#L83) as:

```
#define GRPC_SLICE_MALLOC(len)                                    \
  ((len) <= GRPC_SLICE_INLINED_SIZE                               \
       ? (grpc_slice){.refcount = NULL,                           \
                      .data.inlined = {.length = (uint8_t)(len)}} \
       : grpc_slice_malloc_large((len)))
```

Now `GRPC_SLICE_INLINED_SIZE` is [defined](https://github.com/grpc/grpc/blob/5d0b1aeadecf5cbfafa371da9494cd600615733a/include/grpc/impl/codegen/slice.h#L80) as:

```
#define GRPC_SLICE_INLINED_SIZE (sizeof(size_t) + sizeof(uint8_t *) - 1)
```

If that's 15 in a 64-bit process (my C++ is very rusty) that would explain the difference between "test.Serv.M" (11 bytes) and "test.Serv.Method" (16 bytes).

I'm not *sure* whether that means `grpc_slice_malloc_large` creates a memory leak, but it might...

**Comment by jskeet** [comment]
Aha! I think I've got it. I believe `grpcsharp_channel_create_call` should be calling `grpc_slice_unref(method)` - and potentially the same for `host`.

**Comment by jtattermusch** [comment]
@jskeet I think you're right, thanks for a super detailed analysis.  Once you mentioned "short method names", the first thing I thought of was the difference between by-value and by-ref grpc_slices, but you solved it yourself before I could react.
I'll quickly verify it and submit a patch (it's probably going to v1.4.x branch as we're very close to 1.4.0 release).
